src/content/articles/objects-narrowed-before-usage.mdx

```diff
@@ -0,0 +1,95 @@
+---
+date: "February 17 2023"
+description: "Why TypeScript's type narrowing doesn't apply to object properties that have just been created."
+meta: literals, narrowing, objects, primitives
+---
+
+# Why Object Properties Aren't Narrowed Before Usage
+
+TypeScript's type system boasts a powerful trio of features called _primitives_, _literals_, and _union types_:
+
+- _Primitives_ are types representing the basic immutable value types from JavaScript: `boolean`, `number`, `string`, etc.
+- _Literals_ are types representing specific instances of primitives: `true`, `123`, `"abc"`, etc.
+- _Union types_ are types representing values that could be one of several possible types, such as `string | undefined` representing a value that can be `string` or `undefined`.
+
+> You can think of each _primitive_ as being a _union_ of all the _literals_ of that type.
+
+When you create an object in a TypeScript file, TypeScript's type system needs to figure out what type the object's properties are allowed to be.
+TypeScript will by default use _primitive_ types _(`number`, `string`, ...)_ instead of _literal_ types _(`123`, `"abc"`, ...)_ for properties with a literal value.
+
+For example, `counts.zucchini` in this object is considered to be type `number`, not `2`:
+
+```ts twoslash
+const counts = {
+  apple: 1,
+  zucchini: 2,
+};
+
+counts.zucchini;
+//     ^?
+```
+
+Shouldn't TypeScript know that `counts.zucchini` starts off as the specific literal type `2`, not the general primitive type `number`?
+Can't it tell we haven't changed the object yet?
+
+It can't, but it won't.
+And for very good reason.
+
+<!--truncate-->
+
+## Change Tracking is Impractical
+
+The problem is that it's difficult -oftentimes impossible- for TypeScript to know whether an object has been modified if anything has happened between its initialization and first usage.
+If the object is used in any meaningful application logic, TypeScript won't be able to tell whether it was modified.
+
+### A Practical Example
+
+Let's say you want to `console.log` the `counts` object before reassigning a new value to a count.
+Your use your own custom `log` function that just so happens to only call `console.log`?
```

Comment:
[Nitpicking/proofreading/question] This part seems more like statement. Did you mean to make it a question? 
```suggestion
But your use your own custom `log` function that just so happens to only call `console.log`.
```